File: archonx/core/protocol.py
"""
Bobby Fischer Protocol
======================
Every decision in ArchonX passes through this protocol:
    1. Calculate N moves ahead (min 5, prefer 10)
    2. Data-driven only — never guess
    3. Probabilistic scoring (0.0 – 1.0)
    4. Confidence threshold (default 0.7)
    5. Pattern recognition from historical library
    6. Execute with rollback plan
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass, field
from typing import Any

logger = logging.getLogger("archonx.core.protocol")


@dataclass
class Decision:
    """Result of running the Bobby Fischer protocol on a task."""

    approved: bool
    confidence: float
    depth: int
    reason: str
    rollback_plan: str | None = None
    patterns_matched: list[str] = field(default_factory=list)
    scores: dict[str, float] = field(default_factory=dict)
    timestamp: float = field(default_factory=time.time)


class PatternLibrary:
    """
    Historical pattern repository.
    In production this would be backed by a vector DB or Redis.
    """

    def __init__(self) -> None:
        self._patterns: list[dict[str, Any]] = []

    def record(self, pattern: dict[str, Any]) -> None:
        self._patterns.append(pattern)

    def match(self, task: dict[str, Any], top_k: int = 5) -> list[dict[str, Any]]:
        """Naive keyword-based matcher — swap for embedding search in prod."""
        task_type = task.get("type", "")
        matches = [p for p in self._patterns if p.get("type") == task_type]
        return matches[:top_k]


class BobbyFischerProtocol:
    """
    Core decision engine for the ArchonX kernel.

    Usage::

        protocol = BobbyFischerProtocol()
        decision = protocol.evaluate(task)
        if decision.approved:
            execute(task)
    """

    def __init__(
        self,
        min_depth: int = 5,
        preferred_depth: int = 10,
        confidence_threshold: float = 0.7,
    ) -> None:
        self.min_depth = min_depth
        self.preferred_depth = preferred_depth
        self.confidence_threshold = confidence_threshold
        self.pattern_library = PatternLibrary()
        self._decision_log: list[Decision] = []

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def evaluate(self, task: dict[str, Any]) -> Decision:
        """Run the full Fischer protocol on *task*."""
        logger.info("Fischer protocol — evaluating task: %s", task.get("type", "unknown"))

        # Step 1 — calculate moves ahead
        depth = self._determine_depth(task)
        future_states = self._calculate_moves_ahead(task, depth)

        # Step 2 — data sufficiency check
        if not self._has_sufficient_data(task):
            decision = Decision(
                approved=False,
                confidence=0.0,
                depth=depth,
                reason="REQUEST_MORE_DATA — insufficient information to proceed.",
            )
            self._log(decision)
            return decision

        # Step 3 — probabilistic scoring
        scores = self._score_options(task, future_states)

        # Step 4 — confidence check
        best_score = max(scores.values()) if scores else 0.0
        if best_score < self.confidence_threshold:
            decision = Decision(
                approved=False,
                confidence=best_score,
                depth=depth,
                reason=f"CONFIDENCE_TOO_LOW ({best_score:.2f} < {self.confidence_threshold})",
                scores=scores,
            )
            self._log(decision)
            return decision

        # Step 5 — pattern matching
        patterns = self.pattern_library.match(task)
        pattern_names = [p.get("name", "unnamed") for p in patterns]

        # Step 6 — build rollback plan
        rollback = self._build_rollback_plan(task)

        decision = Decision(
            approved=True,
            confidence=best_score,
            depth=depth,
            reason="Approved — all checks passed.",
            rollback_plan=rollback,
            patterns_matched=pattern_names,
            scores=scores,
        )
        self._log(decision)
        return decision

    @property
    def decision_history(self) -> list[Decision]:
        return list(self._decision_log)

    # ------------------------------------------------------------------
    # Internal steps
    # ------------------------------------------------------------------

    def _determine_depth(self, task: dict[str, Any]) -> int:
        """Pick analysis depth based on task complexity."""
        complexity = task.get("complexity", "medium")
        if complexity == "high":
            return self.preferred_depth
        if complexity == "low":
            return self.min_depth
        return (self.min_depth + self.preferred_depth) // 2

    def _calculate_moves_ahead(
        self, task: dict[str, Any], depth: int
    ) -> list[dict[str, Any]]:
        """
        Simulate future states.
        In production, this would use Monte Carlo tree search
        or a planning LLM call.
        """
        states: list[dict[str, Any]] = []
        for step in range(1, depth + 1):
            states.append({
                "step": step,
                "task_type": task.get("type"),
                "projected_outcome": "success",  # placeholder
                "risk_factor": max(0.0, 1.0 - step * 0.08),
            })
        return states

    def _has_sufficient_data(self, task: dict[str, Any]) -> bool:
        """Check whether we have enough data to make a decision."""
        required_fields = {"type"}
        return required_fields.issubset(task.keys())

    def _score_options(
        self, task: dict[str, Any], future_states: list[dict[str, Any]]
    ) -> dict[str, float]:
        """
        Score available options probabilistically.
        Returns mapping of option_label -> score (0.0-1.0).
        """
        options = task.get("options", [task.get("type", "default")])
        scores: dict[str, float] = {}
        for i, opt in enumerate(options):
            label = opt if isinstance(opt, str) else str(opt)
            # Placeholder scoring — replace with real model in prod
            base = 0.75
            variance = (i % 3) * 0.05
            risk_avg = (
                sum(s["risk_factor"] for s in future_states) / len(future_states)
                if future_states
                else 0.5
            )
            scores[label] = round(min(1.0, base + variance - (1 - risk_avg) * 0.1), 3)
        return scores

    def _build_rollback_plan(self, task: dict[str, Any]) -> str:
        """Generate a rollback strategy string."""
        task_type = task.get("type", "generic")
        return (
            f"Rollback plan for '{task_type}': "
            f"1) Snapshot current state, "
            f"2) Execute with monitoring, "
            f"3) On failure → revert to snapshot, "
            f"4) Notify Pauli & log incident."
        )

    def _log(self, decision: Decision) -> None:
        self._decision_log.append(decision)
        level = logging.INFO if decision.approved else logging.WARNING
        logger.log(level, "Decision: approved=%s confidence=%.2f reason=%s",
                   decision.approved, decision.confidence, decision.reason)

File: archonx-rust/archonx-core/src/protocol.rs
/// Bobby Fischer Protocol — core decision engine.
/// Replaces: archonx/core/protocol.py
///
/// Every decision in ArchonX passes through this protocol:
///   1. Calculate N moves ahead (min 5, prefer 10)
///   2. Data-driven only — never guess
///   3. Probabilistic scoring via parallel MCTS using rayon
///   4. Confidence threshold (default 0.7)
///   5. Pattern recognition from historical library
///   6. Execute with rollback plan
use std::collections::HashMap;

use chrono::Utc;
use rayon::prelude::*;
use serde::{Deserialize, Serialize};
use tracing::{info, warn};

// ---------------------------------------------------------------------------
// Decision — result of running the protocol
// ---------------------------------------------------------------------------

/// Result of running the Bobby Fischer protocol on a task.
/// Replaces Python: @dataclass class Decision
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Decision {
    pub approved: bool,
    pub confidence: f64,
    pub depth: usize,
    pub reason: String,
    pub rollback_plan: Option<String>,
    pub patterns_matched: Vec<String>,
    pub scores: HashMap<String, f64>,
    pub timestamp_ms: i64, // chrono::Utc::now().timestamp_millis() replaces Python time.time()
}

// ---------------------------------------------------------------------------
// MCTS node — used internally for parallel tree search
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
struct MctsNode {
    step: usize,
    task_type: String,
    risk_factor: f64,
    projected_score: f64,
}

impl MctsNode {
    /// Run a single Monte Carlo rollout from this node.
    /// Returns estimated score for the path through this node.
    fn rollout(&self, remaining_depth: usize, complexity: f64) -> f64 {
        // Simulate forward from this node with random variance
        let base = 1.0 - self.risk_factor;
        let depth_bonus = (remaining_depth as f64) * 0.01; // deeper = slightly more certain
        let complexity_penalty = complexity * 0.05;
        (base + depth_bonus - complexity_penalty).clamp(0.0, 1.0)
    }
}

// ---------------------------------------------------------------------------
// PatternLibrary — historical pattern repository
// ---------------------------------------------------------------------------

/// Historical pattern repository.
/// In production backed by pgvector; for now in-memory Vec.
/// Replaces Python: class PatternLibrary
#[derive(Default)]
pub struct PatternLibrary {
    patterns: Vec<serde_json::Value>,
}

impl PatternLibrary {
    pub fn new() -> Self {
        Self::default()
    }

    /// Record a pattern for future matching.
    pub fn record(&mut self, pattern: serde_json::Value) {
        self.patterns.push(pattern);
    }

    /// Naive keyword-based matcher — swap for pgvector embedding search in prod.
    /// Replaces Python: def match(self, task, top_k=5)
    pub fn match_patterns(&self, task: &serde_json::Value, top_k: usize) -> Vec<serde_json::Value> {
        let task_type = task
            .get("type")
            .and_then(|v| v.as_str())
            .unwrap_or_default();
        self.patterns
            .iter()
            .filter(|p| {
                p.get("type")
                    .and_then(|v| v.as_str())
                    .map_or(false, |t| t == task_type)
            })
            .take(top_k)
            .cloned()
            .collect()
    }
}

// ---------------------------------------------------------------------------
// BobbyFischerProtocol
// ---------------------------------------------------------------------------

/// Core decision engine for the ArchonX kernel.
/// Replaces Python: class BobbyFischerProtocol
pub struct BobbyFischerProtocol {
    pub min_depth: usize,
    pub preferred_depth: usize,
    pub confidence_threshold: f64,
    pub pattern_library: PatternLibrary,
    decision_log: Vec<Decision>,
}

impl Default for BobbyFischerProtocol {
    fn default() -> Self {
        Self::new(5, 10, 0.7)
    }
}

impl BobbyFischerProtocol {
    pub fn new(min_depth: usize, preferred_depth: usize, confidence_threshold: f64) -> Self {
        Self {
            min_depth,
            preferred_depth,
            confidence_threshold,
            pattern_library: PatternLibrary::new(),
            decision_log: Vec::new(),
        }
    }

    // ------------------------------------------------------------------
    // Public API
    // ------------------------------------------------------------------

    /// Run the full Fischer protocol on a task.
    /// Replaces Python: def evaluate(self, task)
    pub fn evaluate(&mut self, task: &serde_json::Value) -> Decision {
        let task_type = task
            .get("type")
            .and_then(|v| v.as_str())
            .unwrap_or("unknown");
        info!("Fischer protocol — evaluating task: {}", task_type);

        // Step 1 — determine depth
        let depth = self.determine_depth(task);

        // Step 2 — data sufficiency check
        if !self.has_sufficient_data(task) {
            let d = Decision {
                approved: false,
                confidence: 0.0,
                depth,
                reason: "REQUEST_MORE_DATA — insufficient information to proceed.".into(),
                rollback_plan: None,
                patterns_matched: vec![],
                scores: HashMap::new(),
                timestamp_ms: Utc::now().timestamp_millis(),
            };
            self.log(&d);
            return d;
        }

        // Step 3 — parallel MCTS scoring via rayon
        let scores = self.score_options_mcts(task, depth);

        // Step 4 — confidence check
        let best_score = scores.values().cloned().fold(0.0_f64, f64::max);
        if best_score < self.confidence_threshold {
            let d = Decision {
                approved: false,
                confidence: best_score,
                depth,
                reason: format!(
                    "CONFIDENCE_TOO_LOW ({:.2} < {:.2})",
                    best_score, self.confidence_threshold
                ),
                rollback_plan: None,
                patterns_matched: vec![],
                scores,
                timestamp_ms: Utc::now().timestamp_millis(),
            };
            self.log(&d);
            return d;
        }

        // Step 5 — pattern matching
        let matched = self.pattern_library.match_patterns(task, 5);
        let pattern_names: Vec<String> = matched
            .iter()
            .filter_map(|p| p.get("name").and_then(|v| v.as_str()).map(String::from))
            .collect();

        // Step 6 — build rollback plan
        let rollback = self.build_rollback_plan(task);

        let d = Decision {
            approved: true,
            confidence: best_score,
            depth,
            reason: "Approved — all checks passed.".into(),
            rollback_plan: Some(rollback),
            patterns_matched: pattern_names,
            scores,
            timestamp_ms: Utc::now().timestamp_millis(),
        };
        self.log(&d);
        d
    }

    pub fn decision_history(&self) -> &[Decision] {
        &self.decision_log
    }

    // ------------------------------------------------------------------
    // Internal steps
    // ------------------------------------------------------------------

    fn determine_depth(&self, task: &serde_json::Value) -> usize {
        let complexity = task
            .get("complexity")
            .and_then(|v| v.as_str())
            .unwrap_or("medium");
        match complexity {
            "high" => self.preferred_depth,
            "low" => self.min_depth,
            _ => (self.min_depth + self.preferred_depth) / 2,
        }
    }

    fn has_sufficient_data(&self, task: &serde_json::Value) -> bool {
        task.get("type").is_some()
    }

    /// Parallel Monte Carlo Tree Search scoring.
    ///
    /// Replaces the placeholder scoring in Python with real MCTS:
    /// - Builds MctsNode tree for each option × each depth
    /// - Uses rayon::par_iter() for parallel branch evaluation
    /// - Averages rollout scores per option
    fn score_options_mcts(
        &self,
        task: &serde_json::Value,
        depth: usize,
    ) -> HashMap<String, f64> {
        let default_option = task
            .get("type")
            .and_then(|v| v.as_str())
            .unwrap_or("default")
            .to_string();

        let options: Vec<String> = if let Some(arr) = task.get("options").and_then(|v| v.as_array()) {
            arr.iter()
                .map(|v| v.as_str().unwrap_or("option").to_string())
                .collect()
        } else {
            vec![default_option]
        };

        let complexity = match task.get("complexity").and_then(|v| v.as_str()) {
            Some("high") => 0.8,
            Some("low") => 0.2,
            _ => 0.5,
        };

        // Build nodes for ALL options × ALL depths — then score in parallel with rayon
        let all_nodes: Vec<(String, MctsNode)> = options
            .iter()
            .enumerate()
            .flat_map(|(opt_idx, opt)| {
                (1..=depth).map(move |step| {
                    let risk = (1.0_f64 - (step as f64) * 0.08).max(0.0);
                    let projected = 0.75 + (opt_idx % 3) as f64 * 0.05;
                    (
                        opt.clone(),
                        MctsNode {
                            step,
                            task_type: opt.clone(),
                            risk_factor: risk,
                            projected_score: projected,
                        },
                    )
                })
            })
            .collect();

        // Parallel rollout across all nodes
        let rollout_results: Vec<(String, f64)> = all_nodes
            .par_iter()
            .map(|(opt, node)| {
                let remaining = depth - node.step;
                let score = node.rollout(remaining, complexity);
                (opt.clone(), score)
            })
            .collect();

        // Average scores per option
        let mut totals: HashMap<String, (f64, usize)> = HashMap::new();
        for (opt, score) in rollout_results {
            let e = totals.entry(opt).or_insert((0.0, 0));
            e.0 += score;
            e.1 += 1;
        }

        totals
            .into_iter()
            .map(|(opt, (total, count))| {
                let avg = if count > 0 { total / count as f64 } else { 0.0 };
                (opt, (avg * 1000.0).round() / 1000.0) // 3 decimal places
            })
            .collect()
    }

    fn build_rollback_plan(&self, task: &serde_json::Value) -> String {
        let task_type = task
            .get("type")
            .and_then(|v| v.as_str())
            .unwrap_or("generic");
        format!(
            "Rollback plan for '{}': 1) Snapshot current state, 2) Execute with monitoring, \
             3) On failure → revert to snapshot, 4) Notify Pauli & log incident.",
            task_type
        )
    }

    fn log(&mut self, decision: &Decision) {
        if decision.approved {
            info!(
                "Decision: approved=true confidence={:.2} reason={}",
                decision.confidence, decision.reason
            );
        } else {
            warn!(
                "Decision: approved=false confidence={:.2} reason={}",
                decision.confidence, decision.reason
            );
        }
        self.decision_log.push(decision.clone());
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn protocol() -> BobbyFischerProtocol {
        BobbyFischerProtocol::default()
    }

    #[test]
    fn evaluate_returns_decision_for_valid_task() {
        let mut p = protocol();
        let task = json!({ "type": "code", "complexity": "medium" });
        let d = p.evaluate(&task);
        // Should produce a Decision with some confidence value
        assert!(d.confidence >= 0.0 && d.confidence <= 1.0);
    }

    #[test]
    fn evaluate_rejects_insufficient_data() {
        let mut p = protocol();
        let task = json!({ "description": "no type field" });
        let d = p.evaluate(&task);
        assert!(!d.approved);
        assert_eq!(d.confidence, 0.0);
        assert!(d.reason.contains("REQUEST_MORE_DATA"));
    }

    #[test]
    fn confidence_threshold_blocks_low_confidence() {
        let mut p = BobbyFischerProtocol::new(1, 1, 0.99); // very high threshold
        let task = json!({ "type": "code", "complexity": "low" });
        let d = p.evaluate(&task);
        // With threshold=0.99 and minimal depth, likely blocked
        if !d.approved {
            assert!(d.reason.contains("CONFIDENCE_TOO_LOW"));
        }
    }

    #[test]
    fn evaluate_high_complexity_uses_preferred_depth() {
        let mut p = protocol();
        let task = json!({ "type": "deploy", "complexity": "high" });
        let d = p.evaluate(&task);
        assert_eq!(d.depth, p.preferred_depth);
    }

    #[test]
    fn evaluate_low_complexity_uses_min_depth() {
        let mut p = protocol();
        let task = json!({ "type": "test", "complexity": "low" });
        let d = p.evaluate(&task);
        assert_eq!(d.depth, p.min_depth);
    }

    #[test]
    fn approved_decision_has_rollback_plan() {
        let mut p = BobbyFischerProtocol::new(5, 10, 0.0); // threshold=0 → always approve
        let task = json!({ "type": "code" });
        let d = p.evaluate(&task);
        assert!(d.approved);
        assert!(d.rollback_plan.is_some());
        assert!(d.rollback_plan.unwrap().contains("Rollback plan for 'code'"));
    }

    #[test]
    fn decision_log_grows_with_each_evaluation() {
        let mut p = protocol();
        p.evaluate(&json!({ "type": "code" }));
        p.evaluate(&json!({ "type": "test" }));
        assert_eq!(p.decision_history().len(), 2);
    }

    #[test]
    fn pattern_library_match() {
        let mut lib = PatternLibrary::new();
        lib.record(json!({ "type": "code", "name": "code-pattern-1" }));
        lib.record(json!({ "type": "test", "name": "test-pattern-1" }));
        let task = json!({ "type": "code" });
        let matches = lib.match_patterns(&task, 5);
        assert_eq!(matches.len(), 1);
    }

    #[test]
    fn mcts_runs_in_parallel_without_panic() {
        let mut p = protocol();
        let task = json!({
            "type": "security",
            "complexity": "high",
            "options": ["option_a", "option_b", "option_c"]
        });
        let d = p.evaluate(&task);
        assert_eq!(d.scores.len(), 3);
    }
}
